python/lsst/ctrl/mpexec/cmdLineFwk.py

```diff
@@ -351,6 +353,30 @@ def makeButlerAndCollections(cls, args: SimpleNamespace) -> Tuple[Butler, Sequen
         _LOG.debug("Preparing registry to read from %s and expect future writes to '%s'.", inputs, run)
         return butler, inputs, run
 
+    @staticmethod
+    def defineDatastoreCache() -> None:
+        """Define where datastore cache directories should be found.
+
+        Notes
+        -----
+        All the jobs should share a datastore cache if applicable. This
+        method determines a name and sets an environment variable that
+        will be picked up by the caching class. A value will not be set
+        if one is already defined.
+        """
+        if os.environ.get("DAF_BUTLER_CACHE_DIRECTORY_IF_UNSET"):
+            # A value has already been set.
+            return
+
+        # Come up with a cache name. Create it and immediately delete
+        # it -- there is no nice way to determine a name without creating
+        # it (understandably, this can cause a race condition).
+        cache_dir = tempfile.mkdtemp(prefix="ctrl-mpexec-datastore-cache-")
+        os.rmdir(cache_dir)
```

Comment:
What happens if you don't remove it here?

Reply:
It *should* be automatically deleted at the end regardless. The directory is created during processing only if a cache is created and currently a cache is created only if a file datastore is talking to an object store. Ideally I wouldn't create a directory at all since it won't be needed for a local filesystem datastore and it seems bad to always create it. mkdtemp doesn't let me just come up with a name, so the other option is to create the random name myself and not use mkdtemp at all.

Reply:
Well, I was thinking about uniqueness guarantee of `mkdtemp`. My understanding is that can provide that guarantee because it checks that directory is not already there. If you remove directory but continue to use its name, then another call to `mkdtemp` may create the same directory again?

Reply:
If there is a clash the main hilarity will be another `pipetask` deleting the cache directory whilst the other process is using it.  I have no means of asking whether a cache directory would be needed or not since it all depends on how that specific butler was configured -- the only way to really know that is an explicit datastore API. If I make my own temp name with 20 random letters in it it seems unlikely that there would ever really be a clash and I can check that it wasn't already in use.

Reply:
Okay. I've just pushed a change that includes the process ID and 16 character random string in the temp dir name and I'm hoping that this is going to have a fairly low chance of clashing.